### PR TITLE
fish: 3.0.2 -> 3.1.0

### LINKED ISSUE #79934: Request: upgrade fish 3.0.2 -> 3.1.0
Paging @ocharles

### PR DESCRIPTION
**Pull request by cole-h**
###### Motivation for this change

fish-shell 3.1.0 was just released today with many niceties (including
the ability to have bash-like temporary env vars e.g. `VAR="var1"
command` instead of needing to use `env VAR="var1" command`). To see the
full list of changes, please visit
https://github.com/fish-shell/fish-shell/releases/tag/3.1.0.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc: @ocharles

---

Closes https://github.com/NixOS/nixpkgs/issues/79934.

### CODE REVIEW

**Review by ocharles — APPROVED** [review]
LGTM!

**Comment by mmahut** [comment]
@GrahamcOfBorg build fish

**Comment by madjar** [comment]
The build seems to be failing of osx (https://logs.nix.ci/?key=nixos/nixpkgs.79941&attempt_id=07826efa-a7b1-4528-99ed-5345aba1fb30) with
```
In file included from /tmp/nix-build-fish-3.1.0.drv-0/fish-3.1.0/src/expand.cpp:20:
/nix/store/0xqs5h1n9avficlxbhh299bphpmnpbhk-Libsystem-osx-10.12.6/include/sys/proc.h:119:19: error: field has incomplete type 'struct itimerval'
        struct  itimerval p_realtimer;  /* Alarm timer. */
                          ^
/nix/store/0xqs5h1n9avficlxbhh299bphpmnpbhk-Libsystem-osx-10.12.6/include/sys/proc.h:119:9: note: forward declaration of 'itimerval'
        struct  itimerval p_realtimer;  /* Alarm timer. */
                ^
```

I found nothing relevant in the fish issues, nor in patch changing the version in homebrew. I'm not sure how to proceed.

**Comment by cole-h** [comment]
Unfortunately, I don't have access to any Apple devices. If someone who uses `darwin` could play with this to try and find out what's happening, that would be fantastic. Anybody interested in bisecting it: I can post my `default.nix` I used to bisect another issue previously, but I won't have access to it until this evening.

**Comment by lilyball** [comment]
Interestingly, `sys/proc.h` on my macOS 10.15 system has `#include <sys/time.h>` (`struct itimerval` is defined in that header) but `/nix/store/0xqs5h1n9avficlxbhh299bphpmnpbhk-Libsystem-osx-10.12.6/include/sys/proc.h` is missing that include. That missing include is the proximate cause for the error, so perhaps we can work around it by inserting an `#include <sys/time.h>` at the right spot, but I'll continue investigating.

**Comment by lilyball** [comment]
The failure was introduced by fc0c39b6fdb685ee64121ad58aa68194b3b35588, which removed an import of `<sys/sysctl.h>`, which was indirectly importing `<sys/time.h>`.

**Comment by lilyball** [comment]
I've filled a PR with Fish as https://github.com/fish-shell/fish-shell/pull/6602. We should be able to just fetch and apply https://github.com/fish-shell/fish-shell/pull/6602.patch as a patch.

**Comment by lilyball** [comment]
Well, I suppose fetching the PR as a patch opens us up to hash changes. I'm not sure what the best solution is prior to the PR being merged; we could fetch the commit, but then if the PR's history is edited the commit isn't guaranteed to stay.

Or we could just copy the patch locally.

**Comment by cole-h** [comment]
I'll copy it locally, just to be safe. Thanks for tracking it down!

File: pkgs/shells/fish/default.nix
Comment (lilyball) [inline pkgs/shells/fish/default.nix]: It looks like `__fish_complete_subcommand_root.fish` got effectively inlined into `share/completions/doas.fish` and `share/completions/sudo.fish`, so instead of deleting this, we should retain it but run it on those two files instead.

Though looking further, those files also gained an inclusion of `/usr/local/sbin` as well. Or more specifically, both files have the following line:

```fish
set -lx -a PATH /usr/local/sbin /sbin /usr/sbin
```

I admit I'm not clear as to why these completions are modifying the PATH, or why we're then un-modifying it.

**Comment by madjar** [comment]
I can confirm that this fixes the build on OSX.

File: pkgs/shells/fish/default.nix
Comment (cole-h) [reply pkgs/shells/fish/default.nix]: The commit that originally introduced that behavior was https://github.com/NixOS/nixpkgs/commit/5cffa00443e222c6c56c346de96734b08f2fa060,
which seems to have been working around hardcoded binaries. Nowadays, it appears
that these places use `command` to test the existence of the binary before
executing (though I didn't check every file). I did two things to test whether
or not this find-replace of binaries is still necessary (not foolproof, but
satisfied my curiosity):

* spun up a NixOS VM and overlayed `fish` to have an empty `postInstall`
* ran `fish` inside of `nix-shell --pure` with an empty `postInstall` and a
temporary `HOME` (`nix-shell -p fish --pure` and then `sh -c 'env HOME=$(mktemp
-d) fish'`

and noticed that in neither scenario did I observe `fish` complain about
anything. I ran the following commands from the root of nixpkgs without any
issue:

* `type _`
* `_ asdf`
* `ls`
* `ls | grep pkgs`

We would probably still want the `etcConfigAppendixText` and `fishPreInitHooks`
to be written out in `postInstall`, but everything else appears unnecessary.

I could remove all of the dead `sed` calls and add whatever new ones are
necessary, or I could wipe out `postInstall` except for `etcConfigAppendixText`
and `fishPreInitHooks`.

cc @ocharles -- as the maintainer, what would you prefer?

Comment (lilyball) [reply pkgs/shells/fish/default.nix]: This particular sed in question is mucking with the completions for `sudo` and `doas`. I wish I knew why Fish thought it had to muck with the `PATH` in general for completions of those two commands.

**Comment by mmahut** [comment]
@GrahamcOfBorg build fish

File: pkgs/shells/fish/default.nix
Comment (mmahut) [inline pkgs/shells/fish/default.nix]: Can we please use ```fetchpatch``` from https://github.com/fish-shell/fish-shell/pull/6602 instead of adding the patch in nixpkgs?

Comment (cole-h) [reply pkgs/shells/fish/default.nix]: The only reason I went with adding the patch is because of above:

> Well, I suppose fetching the PR as a patch opens us up to hash changes. I'm not sure what the best solution is prior to the PR being merged; we could fetch the commit, but then if the PR's history is edited the commit isn't guaranteed to stay.

If you still think `fetchpatch` should be used, I'll do so.

File: pkgs/shells/fish/default.nix
Comment (cole-h) [reply pkgs/shells/fish/default.nix]: That is strange. Especially since overwriting with the older `sudo` completion without the `PATH` fiddling, everything works as expected on my system (same amount of suggested binaries). I'll follow your suggestion and `sed` those as well, and open a follow-up PR addressing the necessity of `postInstall`, so we can get this merged sooner rather than later.

Something to be done at a later date is ask upstream whether or not this `PATH` fiddling is really necessary.

File: pkgs/shells/fish/default.nix
Comment (cole-h) [reply pkgs/shells/fish/default.nix]: I decided to `fetchpatch` the commit itself (should hopefully be more static).

**Comment by cole-h** [comment]
Addressed reviews, and added a few spaces to align stuff (I couldn't help myself).

File: pkgs/shells/fish/default.nix
Comment (lilyball) [inline pkgs/shells/fish/default.nix]: As previously mentioned the line was modified in 3.1.0 to be

```fish
set -lx -a PATH /usr/local/sbin /sbin /usr/sbin
```

So this is going to leave it as

```fish
set -lx -a PATH /usr/local/sbin
```

To that end, this should be modified as

```suggestion
      sed -i "s|/usr/local/sbin /sbin /usr/sbin||" \
```

Comment (cole-h) [reply pkgs/shells/fish/default.nix]: Would it be a better idea to just `sed -i "|/usr/local/sbin /sbin /usr/sbin|d"` (delete the matching line) then? No reason to keep it around if it doesn't do anything...

Comment (cole-h) [reply pkgs/shells/fish/default.nix]: After some testing, I realized that `set -lx -a PATH` is a noop, so I'll just go with your suggestion for now.

File: pkgs/shells/fish/default.nix
Comment (lilyball) [reply pkgs/shells/fish/default.nix]: The PR has been merged (and is slated for 3.1.1) so fetching the commit is definitely safe (fetching the PR is safe at this point too, but fetching the commit is just as good).

**Review by lilyball — APPROVED** [review]
LGTM. This passes `nix-review` on darwin

```
https://github.com/NixOS/nixpkgs/pull/79941
1 package built:
fish
```

**Comment by anrddh** [comment]
What additional work needs to be done before this can be merged?

**Comment by cole-h** [comment]
I believe we're just waiting on @ocharles to re-review, since he's the maintainer.

If you can't wait for that, an overlay works just fine (and is what I'm using):

```nix
# ~/.config/nixpkgs/overlays/fish.nix
self: super:

{
  fish = super.fish.overrideAttrs (old: rec {
    version = "3.1.0";

    patches = [ ];

    src = super.fetchurl {
      url =
        "https://github.com/fish-shell/fish-shell/releases/download/${version}/${old.pname}-${version}.tar.gz";
      sha256 = "0s2356mlx7fp9kgqgw91lm5ds2i9iq9hq071fbqmcp3875l1xnz5";
    };

    postInstall = "";
  });
}
```

**Review by ocharles — APPROVED** [review]
This all LGTM, thanks for your hard work!

I should probably step down from maintainership, as you can see I'm not really doing much... maintaining. Anyone interested in stepping up?

**Comment by madjar** [comment]
I can't really maintain this since I'm not using it yet in my weird hybrid osx setup, but I can definitely merge it!

**Comment by terlar** [comment]
Seems this broke the nixos module `programs.fish` (which applies a patch to the fish src):
https://github.com/NixOS/nixpkgs/issues/80423

**Pull request merged**
